sim/paladin/soc.go

```diff
@@ -0,0 +1,168 @@
+package paladin
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/wowsims/sod/sim/core"
+)
+
+// Seal of Command is a spell consisting of:
+// - A judgement that has a flat damage roll, and scales with spellpower.
+// - A 7ppm on-hit proc with a 1s ICD that deals 70% weapon damage and scales with spellpower.
+
+// Judgement of Command has some unusual behaviour in classic:
+// - The judgement operates via a dummy spell, that likely figures out wether or not to apply
+//   half damage if the target is not stunned (counter to the tooltip, the base damage only is
+//   multiplied by 2 if the target is stunned). These dummy spells are implemented as targetting
+//   magic defense type, and have no flags to prevent misses, meaning they roll on spell hit table
+//   and can miss. If it succeeds, it calls the "actual" Judgement of Command spell.
+// - The actual Judgement of Command has flags to not miss and to avoid block/parry/dodge, but
+//   it targets the melee defense type and so crits for double damage.
+// - This is accomplished via the use of the SpellFlagPrimaryJudgement spell flag that is used exclusively by
+//   the base judgement spell. The Seal of Command aura watches for this spell, and casts the actual
+//   Judgement of Command when it successfully is cast.
+
+const socRanks = 5
+
+// Below is the base sp coefficient before it gets reduced by the 70% modifier
+// weapon damage % effect to 20% actual.
+const socProcSpellCoeff = 0.29
+const socJudgeSpellCoeff = 0.423
+
+var socLevels = [socRanks + 1]int{0, 20, 30, 40, 50, 60}
+var socManaCosts = [socRanks + 1]float64{0, 65, 110, 140, 180, 210}
+var socAuraSpellIDs = [socRanks + 1]int32{0, 20375, 20915, 20918, 20919, 20920}
+var socProcSpellIDs = [socRanks + 1]int32{0, 20424, 20944, 20945, 20946, 20947}
+var socJudgeSpellIDs = [socRanks + 1]int32{0, 20467, 20963, 20964, 20965, 20966}
+var socJudgeBasePoints = [socRanks + 1]float64{0, 92, 145, 203, 260, 338}
+var socJudgeRealPointsPerLevel = [socRanks + 1]float64{0, 5.6, 6.1, 5.6, 6.1, 6.1}
+var socEffectDieSides = [socRanks + 1]float64{0, 9, 15, 21, 27, 35}
+var socLevelMinMaxEffects = [socRanks + 1][]int32{{0}, {20, 28}, {30, 38}, {40, 48}, {50, 58}, {60, 60}}
+
+func (paladin *Paladin) applySealOfCommandSpellAndAuraBaseConfig(rank int) {
+	spellIDProc := socProcSpellIDs[rank]
+	spellIDAura := socAuraSpellIDs[rank]
+	spellIDJudge := socJudgeSpellIDs[rank]
+	manaCost := socManaCosts[rank]
+	level := socLevels[rank]
+	scalingLevelMin := socLevelMinMaxEffects[rank][0]
+	scalingLevelMax := socLevelMinMaxEffects[rank][1]
+	judgeBasePoints := socJudgeBasePoints[rank]
+	judgePointsPerLevel := socJudgeRealPointsPerLevel[rank]
+	judgeDieSides := socEffectDieSides[rank]
+
+	levelsToScale := min(paladin.Level, scalingLevelMax) - scalingLevelMin
+	judgeMinDamage := judgeBasePoints + float64(levelsToScale)*judgePointsPerLevel
+	judgeMaxDamage := judgeMinDamage + judgeDieSides
```

Comment:
I notice this in a couple of places now and it seems to be standard for some abilities. Do you think its better to move it into a utility function somewhere in core (maybe spell.go)

Reply:
I agree with that.

When figuring out the damage/damage range on a spell at any particular level, directly inputting the SpellEffect table data and running the relevant calc in the code is imo favourable to the sim implementer manually calculating a min/max at the relevent level brackets.

I've seen at least one instance in the code of a spell's base damage not being properly implemented for the level, having the formula be explicit like this helps in external validation.